README.md
# pimatic-luftdaten

Pimatic plugin retrieves air quality sensor data from Luftdaten.info or a Luftdaten local sensor. Depending on the sensor the following information can be retrieved:
- Particulate Matter (PM 10 and PM 2.5)
- Temperature
- Humidity
- Pressure
- Noise Level (average, max and min levels)

More info at [luftdaten.info](https://luftdaten.info), where you can find instructions for building a luftdaten sensor (incl noise sensor) and how to get de Sensor ID or local IP address. The Air Quality Index (AQI) is calculated from pm10 and pm2.5 sensor data and is based on the [U.S. EPA](https://en.wikipedia.org/wiki/Air_quality_index#United_States).
The plugin supports also the Luftdaten noise sensor. In the gui a dBa classification (0-140 dBa) is added to the noise level data.

The plugin enables automatically attributes when data for an attribute  is received. Sometimes a refresh of the gui is needed to see the values.

For the plugin no API key is required, it uses the open lufdaten.info API or the local sensor API.

### Installation

To install the plugin add the plugin to the config.json of pimatic:
```    
    {
      "plugin": "luftdaten"
    }
```
### Device Configuration

There are 2 devices available

* Luftdaten Device.

Gets air quality data from a specific sensor or area. The Air Quality Index (AQI) data is calculated from that.
Add the device to the devices section:
```    
    {
      "id": "luftdaten",
      "class": "LuftdatenDevice",
      "name": "Device for retrieving data from Luftdaten server",
      "sensorId": "Luftdaten sensor ID or local IP address of sensor",
      "latitude": latitude coordinate of the location you want info for,
      "longitude": longitude coordinate of the location you want info for,
      "radius": the radius of the cel for searching the closest sensor, default 1 km,
      "attributes": [
         "SENSOR_ID", "DISTANCE", "PM10", "PM25",
          "TEMP", "HUM", "BAR", "BAR_SEA",
          "WIFI", "NOISE_LEVEL", "NOISE_LEQ",
          "NOISE_LMIN", "NOISE_LMAX", "AQI",
          "AQI_CODE", "AQI_AIR_QUALITY"
        ] // this is the maximum set of attributes
      "minutes": 60, time between updates
    }
```
You need to configure a Luftdaten SensorId or a latitude/longitude/range combination. If you specify all, the SensorId will be used. If you use de lat/lon option, make sure the range is not to narrow (= no sensors). The maximum radius is 50 km. The device will automatically search for the closest sensors to get a complete data set. In the gui an attribute becomes visible when added in the device config.

* Luftdaten Home Device.

Gets PM10, PM2.5, HUMidity, TEMPerature and BAR data from a local Luftdaten sensor. The Air Quality Index (AQI) data is calculated from that. The Luftdaten device is capable of handling all types of network (Luftdaten cloud and local IP sensor). The Luftdaten Home Device will therefore be removed from the plugin in one of the next versions.

Add the device with the local IP address into the devices section:
```    
    {
      "id": "luftdaten",
      "class": "LuftdatenHomeDevice",
      "name": "Device for retrieving data from local sensor",
      "sensorIp": "local IP address of sensor",
      "minutes": 60, time between updates
    }
```
### Usage

The following variables are available to you in Pimatic for the LuftdatenDevice (after selection in de config!).

* ${luftdaten device id}.PM10             - Particals Matter 10 µg/m³
* ${luftdaten device id}.PM25             - Particals Matter 2.5 µg/m³
* ${luftdaten device id}.TEMP             - Temperature in °C
* ${luftdaten device id}.HUM              - Humidity in %
* ${luftdaten device id}.BAR              - Pressure at the altitude of the sensor in hPa
* ${luftdaten device id}.BAR_SEA          - Pressure at sea level in hPa
* ${luftdaten device id}.WIFI             - Wifi signal strength in dBm
* ${luftdaten device id}.NOISE_LEVEL      - Current noise level classification
* ${luftdaten device id}.NOISE_LEQ        - Current noise level (average per 2.5 minutes) in dBa
* ${luftdaten device id}.NOISE_LMIN       - Lowest noise level in dBa
* ${luftdaten device id}.NOISE_LMAX       - Highest noise level in dBa
* ${luftdaten device id}.AQI              - index between 0 (very good) and 500 (very bad) air quality
* ${luftdaten device id}.AQI_CODE         - color code of air quality (green - purple)
* ${luftdaten device id}.AQI_AIR_QUALITY  - textual info on air quality level
* ${luftdaten device id}.SENSOR_ID        - Luftdaten ID of the sensor
* ${luftdaten device id}.DISTANCE         - Distance to the sensor (km)

The LuftdatenHome Device provides a subset of this data determined by the capabilities of the local sensor.

In the gui an attribute becomes visible when added in the device config.

---------

The plugin is Node v10 compatible and in development. You could backup Pimatic before you are using this plugin!


HISTORY.md
# Release History

* 20190416, V0.2.7
  * Added possibility to use Temperature/Humidity sensor without Pressure (like DHT22). In that case Pressure will have value 0.
* 20190501, V0.3.0
  * Improved API handling and ready for node v8
* 20191109, v0.4.0
  * Luftdaten device is capable of handling Luftdaten and local sensor. In the SensorId field you can input a Luftdaten sensor ID or a local IP address.
  * Luftdaten noise sensor added
  * Added option for latitude, longitude and radius (max 50 km), to find closest sensors
  * Automatic combining of data from closest sensors to get complete dataset
  * Attributes to be shown in the gui, can be selected in the device config
  * The Luftdaten Home device is becoming obsolete and will be removed in a next release.
* 20191109, v0.4.1
  * Changed attribute names back to same names as in previous version (compatibility)
* 20191111, v0.4.2
  * Integrated support files for AQI and DBA calculations
  * All the used sensorId's are now showing in the Gui
* 20191111, v0.4.3
  * bugfix support lib
* 20191120, v0.4.4
  * edit readme.md
* 20191122, v0.4.5
  * better handling of ETIMEDOUT error. After 5 consecutive ETIMEDOUT error's, an error message and reset of values.
* 20191127, v0.4.7
  * bugfix HomeDevice.


dbacalc.js
//'use strict';

var dba_breakpoints = [
    [0, 20],
    [20, 30],
    [30, 50],
    [50, 70],
    [70, 90],
    [90, 120],
    [120, 130],
    [130, 150]
];

var dba_labels = [
    'Silence',
    'Faint',
    'Soft',
    'Moderate',
    'Loud',
    'Very loud',
    'Uncomfortable',
    'Dangerous & painful'
];
var dba_colors = [
    'White',
    'Purple',
    'Blue',
    'Green',
    'Yellow',
    'Red',
    'Brown',
    'Maroon'
];

if (undefined === Array.prototype.findIndex) {
    Array.prototype.findIndex = function (callback) {
        for (var i = 0; i < this.length; i++) {
            if (callback.call(this, this[i], i, this)) {
                return i;
            }
        }
        return -1;
    };
}

function breakpointIndex (value, breakpoints) {
    return breakpoints.findIndex(function (breakpoint) {
        if (null === breakpoint) {
            return false;
        }
        return breakpoint[0] <= value && value <= breakpoint[1];
    });
}

function dba (dbaLevel, breakpoints) {
    var index = breakpointIndex(dbaLevel, breakpoints);
    if (-1 === index) {
        return NaN;
    }
    return index;
}

exports.calc = function (dbaLevel) {
    return dba(dbaLevel, dba_breakpoints);
};

// Airnow.gov descriptions by range
exports.label = function (dbaLevel) {
    var idx = breakpointIndex(dbaLevel, dba_breakpoints);
    return dba_labels[idx];
};

// Aqi color mapping. Returns hex color code.
exports.color = function (dbaLevel) {
    var idx = breakpointIndex(dbaLevel, dba_breakpoints);
    return dba_colors[idx];
};


aqicalc2.js
//'use strict';

var aqi_breakpoints = [
    [0, 50],
    [51, 100],
    [101, 150],
    [151, 200],
    [201, 300],
    [301, 400],
    [401, 500],
];
var pm10_breakpoints = [
    [0, 54],
    [55, 154],
    [155, 254],
    [255, 354],
    [355, 424],
    [425, 504],
    [505, 604],
];
var pm25_breakpoints = [
    [0.0, 12.0],
    [12.1, 35.4],
    [35.5, 55.4],
    [55.5, 150.4],
    [150.5, 250.4],
    [250.5, 350.4],
    [350.5, 500.0],
];
var aqi_labels = [
    'Good',
    'Moderate',
    'Unhealthy for Sensitive Groups',
    'Unhealthy',
    'Very Unhealthy',
    'Hazardous',
    'Hazardous'
];
var aqi_colors = [
    'Green',
    'Yellow',
    'Orange',
    'Red',
    'Purple',
    'Purple',
    'Maroon'
];

if (undefined === Array.prototype.findIndex) {
    Array.prototype.findIndex = function (callback) {
        for (var i = 0; i < this.length; i++) {
            if (callback.call(this, this[i], i, this)) {
                return i;
            }
        }
        return -1;
    };
}

function breakpointIndex (value, breakpoints) {
    return breakpoints.findIndex(function (breakpoint) {
        if (null === breakpoint) {
            return false;
        }
        return breakpoint[0] <= value && value <= breakpoint[1];
    });
}

function aqi (concentration, breakpoints) {
    var index = breakpointIndex(concentration, breakpoints);

    if (-1 === index) {
        return NaN;
    }

    var i_high = aqi_breakpoints[index][1],
        i_low = aqi_breakpoints[index][0],
        c_high = breakpoints[index][1],
        c_low = breakpoints[index][0];

    return Math.round((i_high - i_low) / (c_high - c_low) * (concentration - c_low) + i_low);
}

exports.pm10 = function (concentration) {
    return aqi(concentration, pm10_breakpoints);
};

exports.pm25 = function (concentration) {
    return aqi(concentration, pm25_breakpoints);
};

// Airnow.gov descriptions by range
exports.aqi_label = function (aqi) {
    var idx = breakpointIndex(aqi, aqi_breakpoints);
    return aqi_labels[idx];
};

// Aqi color mapping. Returns hex color code.
exports.aqi_color = function (aqi) {
    var idx = breakpointIndex(aqi, aqi_breakpoints);
    return aqi_colors[idx];
};


lib/calc.js
//'use strict';

var aqi_breakpoints = [
    [0, 50],
    [51, 100],
    [101, 150],
    [151, 200],
    [201, 300],
    [301, 400],
    [401, 500],
];
var pm10_breakpoints = [
    [0, 54],
    [55, 154],
    [155, 254],
    [255, 354],
    [355, 424],
    [425, 504],
    [505, 604],
];
var pm25_breakpoints = [
    [0.0, 12.0],
    [12.1, 35.4],
    [35.5, 55.4],
    [55.5, 150.4],
    [150.5, 250.4],
    [250.5, 350.4],
    [350.5, 500.0],
];
var aqi_labels = [
    'Good',
    'Moderate',
    'Unhealthy for Sensitive Groups',
    'Unhealthy',
    'Very Unhealthy',
    'Hazardous',
    'Hazardous'
];
var aqi_colors = [
    'Green',
    'Yellow',
    'Orange',
    'Red',
    'Purple',
    'Purple',
    'Maroon'
];
var dba_breakpoints = [
    [0, 20],
    [20, 30],
    [30, 50],
    [50, 70],
    [70, 90],
    [90, 120],
    [120, 130],
    [130, 150]
];

var dba_labels = [
    'Silence',
    'Faint',
    'Soft',
    'Moderate',
    'Loud',
    'Very loud',
    'Uncomfortable',
    'Dangerous & painful'
];
var dba_colors = [
    'White',
    'Purple',
    'Blue',
    'Green',
    'Yellow',
    'Red',
    'Brown',
    'Maroon'
];
/*
if (undefined === Array.prototype.findIndex) {
    Array.prototype.findIndex = function (callback) {
        for (var i = 0; i < this.length; i++) {
            if (callback.call(this, this[i], i, this)) {
                return i;
            }
        }
        return -1;
    };
}

function breakpointIndex (value, breakpoints) {
    return breakpoints.findIndex(function (breakpoint) {
        if (null === breakpoint) {
            return false;
        }
        return breakpoint[0] <= value && value <= breakpoint[1];
    });
}
*/

if (undefined === Array.prototype.findIndex) {
    Array.prototype.findIndex = function (callback) {
        for (var i = 0; i < this.length; i++) {
            if (callback.call(this, this[i], i, this)) {
                return i;
            }
        }
        return -1;
    };
}

function breakpointIndex (value, breakpoints) {
    return breakpoints.findIndex(function (breakpoint) {
        if (null === breakpoint) {
            return false;
        }
        return breakpoint[0] <= value && value <= breakpoint[1];
    });
}

function dba (dbaLevel, breakpoints) {
    var index = breakpointIndex(dbaLevel, breakpoints);
    if (-1 === index) {
        return NaN;
    }
    return index;
}

function aqi (concentration, breakpoints) {
    var index = breakpointIndex(concentration, breakpoints);

    if (-1 === index) {
        return NaN;
    }

    var i_high = aqi_breakpoints[index][1],
        i_low = aqi_breakpoints[index][0],
        c_high = breakpoints[index][1],
        c_low = breakpoints[index][0];

    return Math.round((i_high - i_low) / (c_high - c_low) * (concentration - c_low) + i_low);
}


exports.pm10 = function (concentration) {
    return aqi(concentration, pm10_breakpoints);
};

exports.pm25 = function (concentration) {
    return aqi(concentration, pm25_breakpoints);
};

// Airnow.gov descriptions by range
exports.aqi_label = function (aqi) {
    var idx = breakpointIndex(aqi, aqi_breakpoints);
    return aqi_labels[idx];
};

// Aqi color mapping. Returns hex color code.
exports.aqi_color = function (aqi) {
    var idx = breakpointIndex(aqi, aqi_breakpoints);
    return aqi_colors[idx];
};

exports.dba = function (dbaLevel) {
    return dba(dbaLevel, dba_breakpoints);
};

// Airnow.gov descriptions by range
exports.dba_label = function (dbaLevel) {
    var idx = breakpointIndex(dbaLevel, dba_breakpoints);
    return dba_labels[idx];
};

// Aqi color mapping. Returns hex color code.
exports.dba_color = function (dbaLevel) {
    var idx = breakpointIndex(dbaLevel, dba_breakpoints);
    return dba_colors[idx];
};
